### PR TITLE
[WIP] a new criterion for length normalization

### PR DESCRIPTION
**Pull request by shhshn**
Fixes issue #1320 by introducing a new stopping criterion as proposed by Yang et al. (2018).
- http://aclweb.org/anthology/D18-1342

(Note: this is a draft PR based on an old version. See #1344 for the latest version.)

Followings are examples using the [WMT English-to-German model](http://opennmt.net/Models-py/). Only the `-fast` decoding path is supported. Performance may degrade significantly.

- `-fast [-length_penalty none]`
```shell
$ echo "▁28 - Y ear - O ld ▁Chef ▁Found ▁Dead ▁at ▁San ▁Francisco ▁Mal l" | python translate.py -model averaged-10-epoch.pt -src /dev/stdin --output /tmp/t -gpu 0 -fast; cat /t
PRED AVG SCORE: -0.8590, PRED PPL: 2.3608
▁28 - Jahr - O ld ▁Chef ▁Found ▁Dead
```

- `-fast -length_penalty avg` without this PR
```shell
$ echo "▁28 - Y ear - O ld ▁Chef ▁Found ▁Dead ▁at ▁San ▁Francisco ▁Mal l" | python translate.py -model averaged-10-epoch.pt -src /dev/stdin --output /tmp/t -gpu 0 -fast -length_penalty avg; cat /tmp/t
PRED AVG SCORE: -0.8590, PRED PPL: 2.3608
▁28 - Jahr - O ld ▁Chef ▁Found ▁Dead
```

- `-fast -length_penalty avg` with this	PR
```shell
PRED AVG SCORE: -0.0311, PRED PPL: 1.0316
▁28 - Jahr - O ld - C he f ▁To te ▁in ▁San ▁Francisco ▁Mal l
```

### CODE REVIEW

**Comment by yilinyang7** [comment]
Hi! Thank you for your efforts. 
In the original paper, we set "predicted_length" as "max_length", thus the beam search won't go too long. However, to get a "predicted_length", you either need to use a tgt/src length ratio from training set, or (like in the paper) to use a real length_prediction model to predict the tgt length.

Thanks,
Yilin

**Comment by shhshn** [comment]
Hi! Thank you for the clarification. I got your point and will look into it.

I really like your paper in that it gives a concrete explanation for the long-standing length normalization problem and solutions to the problem.

**Comment by yilinyang7** [comment]
Glad you like it :)

**Comment by vince62s** [comment]
@shhshn Thanks for your contribution but your repo is far behind vs master.
You need to pull and redo the PR because the codebase has changed.
It should not be so difficult but most of changes should happen now in beam_search.py
Let me know if you plan to work on this or if we need to pick it up.
Thanks.

**Comment by vince62s** [comment]
Based on @yilinyang7 's comment, I tested this code and changed max_length to memory_lengths[i] divided by an average ratio for EN/DE but it gives me worse scores vs the current "avg".
By the way, in the original post, you can see that the PRED AVG SCORE is too low, something must be wrong.

**Comment by yilinyang7** [comment]
Using a predicted length as max_length might give you lower scores, but it won't render too long generation which will hurt BLEU (while also being inefficient).

**Comment by vince62s** [comment]
okay, after further testing on current master, it seems to give good results if I turn off length_penalty avg.
In other words, if we normalize, it does not seem to work which looks strange.

Also, to better understand, how do you link your formula with the paper sect 5.2 equation(13) ?
thanks.
